Users - base work

pavish opened the pull request:
Related to #2205 
Fixes #2027 
Fixes #2404 

**This PR:**
* Implements base work required for managing users, user profile.
  * User profile preloading and store
  * Users store
  * Routes and pages
  * Users Api utility
* Implements User profile page
* Implements a functional **un-styled** users listing, add, and edit pages
* Has general improvement changes to existing utils

**Note for reviewers:**
* I have not styled the users listing, new user creation, and edit user pages. However, they are functional. The styling portion is beyond the scope of this PR, and I've added task list items to the meta ticket for them.
* Regarding changes to breadcrumbs.
  * Profile page and administration page do not fall into any database.
  * The breadcrumbs are updated in a minimally invasive way, with consideration that we'll have to come back to this when we start supporting multiple DBs.
  * It is currently structured in a way that: clicking on logo within a non db route will go to '/' and within a DB route will go to '/database'.
  * I've removed the `showMinimal` logic since it doesn't seem to be required.
* Users store structure
  * We have a users store which contains the users list, but this is not initialized at the app level.
  * It is initialized at the /users route.
  * This structure wasn't mentioned on the RFC, because it wasn't base work related to the entire feature, but this will require some attention since it introduces a new convention.
* Field now clears their own server side errors when user edits them. 
* User role: `admin, user` is currently not sent in requests. It is blocked by https://github.com/centerofci/mathesar/issues/2392.

**Screenshots:**
### App header setting dropdown menu
  <img width="215" alt="Screenshot 2023-01-31 at 8 46 14 AM" src="https://user-images.githubusercontent.com/11850603/215654130-1625b72f-9cb7-43cf-9265-35dfdcf10dfb.png">

### Profile page
  ![Screenshot 2023-01-31 at 8 48 23 AM](https://user-images.githubusercontent.com/11850603/215654507-bae8261a-ede1-46f8-9170-69987d183682.png)
  * Changes from Figma:
    * Password is separated out into a different section since the requests are different for changing/resetting password, as finalized for the users add/edit page.
    * It contains an old password input, which isn't present in figma, but is required by the API.
    
### User Administration pages - functional but isn't styled
  <img width="625" alt="Screenshot 2023-01-31 at 8 56 28 AM" src="https://user-images.githubusercontent.com/11850603/215655552-fe97aa1e-1c0e-40ef-9ff4-24d5b9e163b0.png">
  <img width="806" alt="Screenshot 2023-01-31 at 8 57 13 AM" src="https://user-images.githubusercontent.com/11850603/215655573-36b00150-b67f-440e-8c20-1dca3dd94c49.png">


## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

seancolsen commented on mathesar_ui/src/stores/userProfile.ts:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  /* @throws Error if unable to save */
+  async update(
```
This feels fishy to me.

- We have classes like `ColumnsDataStore` which are very clearly **mutable** because:
    - Methods are named with verbs like `rename` and `patch`
    - Methods sometimes mutate the instance
    - Methods never return the instance
    - Methods sometimes mutate back-end data too

- We have classes like `Filtering` which are very clearly **immutable** because:
    - Methods are named with nouns and phrases like `withEntry` and `withoutColumns`
    - Methods never mutate the instance 
    - Methods always return a new instance
    - Methods never mutate backend data

- This `UserProfile` class doesn't fit squarely into either of the categories above, which makes me concerned that it may lead to confusion and readability problems. I would consider pulling the API logic out of the class, potentially into a new _mutable_ class called `UserProfileStore` which composes `UserProfile`.

pavish replied:
Thanks. Changes made in https://github.com/centerofci/mathesar/pull/2390/commits/031d821032982b2f174433e82ac3852a79b2e565

pavish replied:
This should also fix the `Function called outside component initialization` error.

Getting the context from out of a component initialization threw this error, I did a final set of refactoring while raising the PR and missed to factor this in.

seancolsen reviewed (commented):
Very cool! Thanks for laying all this groundwork, @pavish. I have some changes to request, but I'd be fine with merging this PR as-is and making them in subsequent PRs. Up to you.

- I added some commits

    - f66d8e3e54e4cd9412863370c0586d55b8b76246 because in `PasswordChangeForm.svelte` you had:

        ```svelte
        <WarningBox fullWidth>
          ...
        </WarningBox>
        ```

        This gives me a type error in my editor because WarningBox doesn't accept a `fullWidth` prop. I've also opened a [front end code standards PR](https://github.com/centerofci/mathesar-wiki/pull/77) proposing that we make this a standard. That PR has some more info in it about my rationale here.

    - Small stuff in 5e9003c21c68d60a1c1a997726ddb05741282535 d487bd6626c06a2026b355f7e473452c11a1a91a

    - In bb7d0d810bcd29b9bd2e3dccace4877eabd0057e I tweaked some code for better type-safety.
    
        I also added 37e8e2de384052719f6ec31a42e7f3bee2a4e1ac, installing the `type-fest` package which provides the `UnionToIntersection` utility type I used, Since this is a dev-only dependency, I figure it's not that big of a deal to add it. I've been noticing we've been needing more complex utility types more regularly, and I though this package would help.

    - In 74fb0f1355c6e5daf23309b82fa497d1b4cdb302 I changed the form validation on the password fields to use "combo validation". Before this change, it was possible for the user to enter a password into the second field, then enter a _different_ password into the first field without triggering any validation errors. Field-specific validators (set directly on the field) only run when the field value changes. Combo validators run when _any_ field in the whole form changes, so that's what we need here. I also added some additional helper combo validation utils functions. This way we can make use of the slightly simpler `comboMustBeEqual` validator which doesn't need a predicate argument.

        It's also worth mentioning that the form system currently displays combo errors on all of the fields used for the combo validation. This UX feels a little weird to me. Ideally you'd see the error message only on the field which you last changed. Or maybe at least only on one field specified within the combo validator function. I'm open to improving this within the form system, but I think it's okay for now.

- When I try to update my user profile I see this weird Svelte error

    > Function called outside component initialization

    ![image](https://user-images.githubusercontent.com/42411/215852786-b79b5290-e059-4436-87aa-e413810cb0aa.png)

- Using PascalCase for `UserApi` (across 4 files) seems weird to me because it's an instance. I'd expect it to be `userApi`.


- I added #2404 requesting that we remove the "Short Name" field until we have a use-case for it. No work needed in this PR, but I just wanted to point this out.

pavish commented on mathesar_ui/src/pages/user-profile/ProfilePage.svelte:
```diff
@@ -0,0 +1,69 @@
+</LayoutWithHeader>
+
+<style>
+  h1 {
```
I agree. We have not established a pattern for this yet and the styling of headers differ from page to page. For eg., table names & exploration names need a different typography compared to the headers in the pages in the above screenshot. 

I think this is out of scope for this PR, and we should handle it separately. I'll create a separate issue for this.

pavish replied:
Ticket: https://github.com/centerofci/mathesar/issues/2416

pavish commented on mathesar_ui/src/component-library/common/icons.ts:
```diff
@@ -54,6 +55,7 @@ export const iconUploadFile: IconProps = { data: faFileUpload };
 
 export const iconFile: IconProps = { data: faFile };
 export const iconSettings: IconProps = { data: faGears };
+export const iconSettingsMajor: IconProps = { data: faGear };
```
Changes made in https://github.com/centerofci/mathesar/pull/2390/commits/e35e2e091cc6859c1e33837d8383b311ccfda7a9

pavish commented on mathesar_ui/src/api/users.ts:
```diff
@@ -0,0 +1,81 @@
+export interface SchemaRole {
+  schema: SchemaEntry['id'];
+  role: UserRole;
+}
```
Thanks. Changes made in https://github.com/centerofci/mathesar/pull/2390/commits/5572f7edea2ccd7a82907023c4a73ffe710c5690.

pavish commented:
> Using PascalCase for UserApi (across 4 files) seems weird to me because it's an instance. I'd expect it to be userApi.

Fixed in https://github.com/centerofci/mathesar/pull/2390/commits/233e95582e8319e8bd413fabef3ef78b91a6a10b.

> I added https://github.com/centerofci/mathesar/issues/2404 requesting that we remove the "Short Name" field until we have a use-case for it.

Changed in  https://github.com/centerofci/mathesar/pull/2390/commits/9ef3b47970d4d90cda7b15ba0d614fab5a997297

pavish commented:
@seancolsen The PR is ready for review again.

seancolsen reviewed (approved):
This looks great now. Even though we have an outstanding review request aimed at @rajatvijay, I'm going ahead and merging this PR anyway since it's been open a while and we need to move fast on this feature.
